source/src/ble/cs_BleCentral.cpp

```diff
@@ -0,0 +1,755 @@
+/*
+ * Author: Crownstone Team
+ * Copyright: Crownstone (https://crownstone.rocks)
+ * Date: Apr 9, 2021
+ * License: LGPLv3+, Apache License 2.0, and/or MIT (triple-licensed)
+ */
+
+#include <ble/cs_BleCentral.h>
+#include <ble/cs_Nordic.h>
+#include <ble/cs_Stack.h>
+#include <common/cs_Types.h>
+#include <events/cs_Event.h>
+#include <events/cs_EventDispatcher.h>
+#include <logging/cs_Logger.h>
+
+const uint16_t WRITE_OVERHEAD = 3;
+const uint16_t LONG_WRITE_OVERHEAD = 5;
+
+void handle_discovery(ble_db_discovery_evt_t* event) {
+	BleCentral::getInstance().onDiscoveryEvent(event);
+}
+
+BleCentral::BleCentral() {
+	_discoveryModule.discovery_in_progress = false;
+	_discoveryModule.discovery_pending = false;
+	_discoveryModule.conn_handle = BLE_CONN_HANDLE_INVALID;
+	uint32_t nrfCode = ble_db_discovery_init(handle_discovery);
```

Comment:
Be aware that because there is a static BleCentral instance, the constructor is executed before `main()` is called. Can this method be called at that time?

Reply:
I don't think this is true. This is using the Meyers singleton pattern, see http://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html and https://stackoverflow.com/questions/1008019/c-singleton-design-pattern which is lazy-evaluated.

Reply:
I think it's constructed at the first .getInstance() call. But I think it's better to move this code to the init() function anyway.